### PR TITLE
Include GL extension fliter/allowlist if and only if its needed

### PR DESCRIPTION
**Pull request by sbc100**
Here we use the JS library dependency mechanism to include this code only when its needed.

Sadly this will end up being present in almost all GL-using programs because most of them will
end up using `emscriptenWebGLGet` (e.g. any app that calls `glGetInteger`, `glGetString`, etc).
The code size impact of this change (for codebases that call glGetString but don't otherwise
call GetProcAddress is 858 bytes of JS or 356 compressed).

Fixes: #20798

### CODE REVIEW

**Comment by sbc100** [comment]
I measured the code size impact of this change for codebases that call glGetString but don't call GetProcAddress as 858 bytes of JS or 356 compressed JS.  @juj,  f you think we need a switch to make this less conservative we could add one perhaps.

File: src/library_webgl.js
Comment (kainino0x) [inline src/library_webgl.js:142]: nit: Update this variable name

File: src/library_webgl.js
Comment (kainino0x) [inline src/library_webgl.js:790]: I wonder if there is any risk to removing this shim - now any user JS code that checks the list of supported extensions (on the context directly) will get a different list than WASM code. I think it is OK though.

**Review by kainino0x — APPROVED** [review]
LGTM

File: src/library_webgl.js
Comment (sbc100) [reply src/library_webgl.js:790]: I don't think user code should be doing anything directly with the GL context though should it?  If it does it certainly should know what its doing.

File: src/library_webgl.js
Comment (sbc100) [reply src/library_webgl.js:142]: Done

File: src/library_webgl.js
Comment (kainino0x) [reply src/library_webgl.js:790]: I think it's not uncommon to use the context inside EM_ASM blocks, for example to use interop entrypoints like `texImage2D(ImageBitmap)`. There are probably other cases of applications mixing JS and WASM code for various reasons, and either grabbing the context out of Emscripten or calling `canvas.getContext('webgl')` which will re-return the same context object. But it wouldn't be so common for such code to be enabling extensions.

Also I forgot this shim is new, and 2 weeks ago the JS-exposed extension list was never filtered, so I don't think there's risk of breaking existing applications. Only risk of confusing developers slightly, I think.

Comment (sbc100) [reply src/library_webgl.js:790]: I also think there is an argument for not filter the list for users that directly go under the hood and access the context.  In that case, such users are more likely to want to get the unfiltered list, since in those circumstances they might not be limited to extensions that emscripten's C++ library supports and they might want to see the real/underlying list

**Review by juj — CHANGES_REQUESTED** [review]
Sorry, I don't agree with this PR. We should not land this, see https://github.com/emscripten-core/emscripten/issues/20798#issuecomment-1865200414 for rationale.

LGTM to land if there is an opt-out for developers who statically know their code will not cause problems.

File: src/library_html5_webgl.js
Comment (juj) [inline src/library_html5_webgl.js:444]: This function `emscripten_webgl_get_supported_extensions()` is called by native WebGL rendering engines (so groups 1 and 2 in https://github.com/emscripten-core/emscripten/issues/20798#issuecomment-1865200414), so this one should not get the allowlisted treatment.

Comment (sbc100) [reply src/library_html5_webgl.js:444]: sgtm!

**Comment by juj** [comment]
Although, won't this still leave this code path unattended? https://github.com/emscripten-core/emscripten/blob/53ce9c759e47ac6acd4c454b360f9f922d81c97d/src/library_webgl.js#L1205-L1215

**Comment by sbc100** [comment]
> Although, won't this still leave this code path unattended?
> 

Ah yes, I misunderstood that point of that callsite and assumed it wanted the unfiltered list

File: src/library_html5_webgl.js
Comment (sbc100) [reply src/library_html5_webgl.js:444]: Done

**Comment by sbc100** [comment]
Hopefully this address all the concern and can finally close #20798.   Will wait for @juj to get back from vacation for final approval.

File: src/library_webgl.js
Comment (kainino0x) [inline src/library_webgl.js:1221]: nit: This naming convention makes glGetExtensions sound like a gl.h entrypoint (except for the `$`), but it isn't.
Hopefully it can be somewhat clear how it's different from getEmscriptenSupportedExtensions. Right now it feels arbitrary which ones is called in which place.

(I'm actually not totally sure why we expose GL_EXTENSIONS entries without prefix - I think OpenGL always prefixes them, and only WebGL, which doesn't even have `gl.getString(gl.EXTENSIONS)`, doesn't. But that's beside the point.)

Maybe
getEmscriptenSupportedWebGLExtensionsUnprefixed
getEmscriptenSupportedWebGLExtensionsPlusPrefixed

("WebGL" distinguishing it from other GL extensions like the ones in glemu.)

File: src/library_glemu.js
Comment (kainino0x) [inline src/library_glemu.js:397]: Since all of the "fake" extensions are GL_ prefixed I'm guessing that here we would want to get GL_ prefixed versions of WebGL extensions (if enabled) too. Though I see that doesn't seem to have been true before.

**Review by kainino0x — COMMENTED** [review]
This is getting a bit deep into things for me. Since Jukka will have to review anyway I will just defer to him. Few small comments though.

File: src/library_webgl.js
Comment (juj) [reply src/library_webgl.js:1221]: > I'm actually not totally sure why we expose GL_EXTENSIONS entries without prefix

The reason for this was:

1. originally from dawn of time, Emscripten exposed GL_EXTENSIONS list without GL_ prefixing the extensions,
2. when working on porting Unreal Engine 3 in 2013, Vlad Vukicevic realized that of course the extensions won't match what UE3 would expect, so he asked me to add GL_EXTENSIONS to advertise the prefixed variants,
3. we didn't dare to remove advertising the original unprefixed variants, at the basis of "maybe that could break existing codebases targeting Emscripten that might be reading the unprefixed extensions".

So we decided to leave them in both forms.

Comment (juj) [reply src/library_webgl.js:1221]: > nit: This naming convention makes glGetExtensions sound like a gl.h entrypoint (except for the $), but it isn't.

Agree with this. With respect to the GL spec, naming custom functions with `glFoo()` is a bit similar to adding new functions in the `std::` namespace in C++.

Not that we'd ever have problems with a function named `glGetExtensions`, though there is some sense of misplacedness to it.

Maybe this could be `$webglGetExtensions` instead?

**Review by juj — APPROVED** [review]
LGTM

File: src/library_webgl.js
Comment (sbc100) [reply src/library_webgl.js]: Done

File: src/library_glemu.js
Comment (sbc100) [reply src/library_glemu.js:397]: Agreed.  Perhaps that should be a followup though, since it seems like a separate change/fix.

**Comment by floooh** [comment]
Could this change have broken the `enableExtensionsByDefault` behaviour of the `emscripten_webgl_create_context()` call? Since emsdk 3.1.52 I'm no longer seeing the ETC2 and ASTC extension advertised (`WEBGL_compressed_texture_etc` and 	`WEBGL_compressed_texture_astc` event though they are supported by WebGL2 here: https://webglreport.com/?v=2

**Comment by kainino0x** [comment]
That is probably the problem. I thought I checked that this list had all the extensions I expected it to have from https://registry.khronos.org/webgl/extensions/, but I may have forgotten to check the community approved extensions in addition to the ratified ones.

**Comment by sbc100** [comment]
> That is probably the problem. I thought I checked that this list had all the extensions I expected it to have from https://registry.khronos.org/webgl/extensions/, but I may have forgotten to check the community approved extensions in addition to the ratified ones.

It this PR in particular?  I assume if there are issues with the contents of the list then #20802 would be the root cause.  This PR was mostly refactoring.

**Comment by kainino0x** [comment]
Right, I didn't check which PR this was. I think it must have been that one.

**Comment by floooh** [comment]
Thanks for looking into this, and apologies too for identifying the wrong PR.

**Pull request merged**
